instrumentation-tests/src/androidTest/java/com/mapbox/navigation/instrumentation_tests/core/HandlerThreadTest.kt

```diff
@@ -0,0 +1,88 @@
+import com.mapbox.navigation.core.MapboxNavigationProvider
+import com.mapbox.navigation.core.internal.extensions.flowOnFinalDestinationArrival
+import com.mapbox.navigation.instrumentation_tests.activity.EmptyTestActivity
+import com.mapbox.navigation.instrumentation_tests.utils.MapboxNavigationRule
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.sdkTest
+import com.mapbox.navigation.instrumentation_tests.utils.coroutines.setNavigationRoutesAsync
+import com.mapbox.navigation.instrumentation_tests.utils.location.MockLocationReplayerRule
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider
+import com.mapbox.navigation.instrumentation_tests.utils.routes.RoutesProvider.toNavigationRoutes
+import com.mapbox.navigation.testing.ui.BaseTest
+import com.mapbox.navigation.testing.ui.utils.getMapboxAccessTokenFromResources
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.android.asCoroutineDispatcher
+import kotlinx.coroutines.flow.first
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNotEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+class HandlerThreadTest : BaseTest<EmptyTestActivity>(EmptyTestActivity::class.java) {
+
+    override fun setupMockLocation(): Location = mockLocationUpdatesRule.generateLocationUpdate {
+        latitude = 38.894721
+        longitude = -77.031991
+    }
+
+    @get:Rule
+    val mockLocationReplayerRule = MockLocationReplayerRule(mockLocationUpdatesRule)
+
+    private lateinit var mapboxNavigation: MapboxNavigation
+    private lateinit var sdkThread: HandlerThread
+    private lateinit var handler: Handler
+    private lateinit var sdkDispatcher: CoroutineDispatcher
+
+    @Before
+    fun setUp() {
+        sdkThread = HandlerThread("test sdk thread")
+        sdkThread.start()
+        handler = Handler(sdkThread.looper)
+        sdkDispatcher = handler.asCoroutineDispatcher()
+        handler.post {
+            mapboxNavigation = MapboxNavigationProvider.create(
+                NavigationOptions.Builder(activity)
+                    .accessToken(getMapboxAccessTokenFromResources(activity))
+                    .build()
+            )
+        }
+    }
+
+    @After
+    fun tearDown() {
+        handler.post {
+            mapboxNavigation.onDestroy()
+        }
+        sdkThread.quitSafely()
+    }
+
+    private fun assertSDKThread() {
+        assertEquals(sdkThread.looper, Looper.myLooper())
+    }
+
+    @Test
+    fun set_navigation_routes_successfully() = sdkTest(dispatcher = sdkDispatcher) {
+        val routes = RoutesProvider.dc_very_short(activity).toNavigationRoutes()
+        mapboxNavigation.setNavigationRoutesAsync(routes)
+        mapboxNavigation.startTripSession()
+        mockLocationReplayerRule.playRoute(routes.first().directionsRoute)
+        var routeProgressEventsCount = 0
+        mapboxNavigation.registerRouteProgressObserver {
```

Comment:
Since the observers are sticky I think it would make sense to wait for several updates: the first one may be from the caller thread (it's very straightforward, it's hard to do it otherwise) but the one that's invoked later (not on registration) - may be invoked from a different thread.

Reply:
agree, I had similar motivation writing this test. Do you think that the check is okay or maybe you have suggestions/improvements?

Reply:
What about:
```
        mapboxNavigation.registerRouteProgressObserver {
            assertSDKThread()
            routeProgressEventsCount++
        }
        mapboxNavigation.routeProgressUpdates().take(2).toList()
```
And you won't need the last assert then.